pkg/services/ngalert/state/manager.go

```diff
@@ -99,11 +110,69 @@ func (st *Manager) Run(ctx context.Context) error {
 		case <-ctx.Done():
 			st.log.Debug("Stopping")
 			ticker.Stop()
+			st.stateRunnerShutdown <- struct{}{}
+			st.stateRunnerWG.Wait()
 			return ctx.Err()
 		}
 	}
 }
 
+func (st *Manager) startSync(ctx context.Context) {
+	st.stateRunnerWG.Add(1)
+	go func(ctx context.Context) {
+		ticker := st.clock.Ticker(time.Second * 30)
```

Comment:
We might want to increase this to something like 5 minutes or even make it configurable. As we always save on a clean shutdown, I think 5 minutes might be a good start. Any thoughts?

Reply:
I would agree to making it configurable, and I think 5 minutes is a more reasonable default, as if you need this feature, it's because saving states is taking too long.

Reply:
Agreed, this should be configurable and I would say go as high as 10 minutes as a default. In my opinion, adequate performance under regular operations is more important than precise recovery on unclean shutdown for alerting state.

Reply:
Instead of having a static configuration option that we will probably have to set for many different stacks in Hosted Grafana (as we have very varied workloads), should the timer be computed as a weighted average of the past N durations?

For example:

```
<-time.After(min(30 * time.Second), weightedAverage())
```

For example, if the last N syncs took 15 minutes, then wait 15 minutes between syncs. If it took 1 minute, wait 1 minute between syncs. We apply a minimum wait of 30 seconds. This just avoids a situation where we are always catching up and writing to the database in what is effectively a tight loop for large workloads.

Reply:
Instead of having a static configuration option that we will probably have to set for many different stacks in Hosted Grafana (as we have very varied workloads), should the timer be computed as a weighted average of the past N durations?

Perhaps something like the following:

```
<-time.After(min(30 * time.Second), weightedAverage())
```

For example, if the last N syncs took 15 minutes, then wait 15 minutes between syncs. If it took 1 minute, wait 1 minute between syncs. We apply a minimum wait of 30 seconds. This just avoids a situation where we are always catching up and writing to the database in what is effectively a tight loop for large workloads.